### PR TITLE
feat(storage): use roaring bitmaps for history tables

### PR DESCRIPTION
**Pull request by shekhirin**
## Performance benchmarks

X axis is the random delta in the range of `x-x/2..=x+x/2` between 2000 integers

image
image

## Size benchmarks

Mainnet Archive Node
|                  | Elias-Fano | Roaring Bitmaps |
|------------------|------------|-----------------|
| `AccountHistory` | 63 GB    | 37 GB         |
| `StorageHistory` | 389 GB   | 220 GB        |

Sepolia Full Node
|                  | Elias-Fano | Roaring Bitmaps |
|------------------|------------|-----------------|
| `AccountHistory` | 82 MB    | 23 MB         |
| `StorageHistory` | 712 MB   | 223 MB        |

### CODE REVIEW

**Comment by shekhirin** [comment]
Sepolia Full Node

|                  | Elias-Fano | Roaring Bitmaps |
|------------------|------------|-----------------|
| `AccountHistory` | 82 MB    | 23 MB         |
| `StorageHistory` | 712 MB   | 223 MB        |

---

Pending Mainnet sync

File: crates/storage/provider/src/providers/state/historical.rs [hunk 1]
```diff
@@ -110,10 +111,16 @@ impl<'b, TX: DbTx> HistoricalStateProviderRef<'b, TX> {
...
         if let Some(chunk) = cursor.seek(key)?.filter(|(key, _)| key_filter(key)).map(|x| x.1 .0) {
-            let chunk = chunk.enable_rank();
+            // Get the rank of the first entry before or equal to our block.
+            let mut rank = chunk.rank(self.block_number);
 
-            // Get the rank of the first entry after our block.
-            let rank = chunk.rank(self.block_number as usize);
+            // Adjust the rank, so that we have the rank of the first entry strictly before our
+            // block (not equal to it).
+            if rank.checked_sub(1).and_then(|rank| chunk.select(rank)) == Some(self.block_number) {
```
Comment (joshieDo) [inline crates/storage/provider/src/providers/state/historical.rs:119]: til

**Review by joshieDo — APPROVED** [review]
faster, smaller lfg. i'd just remove any mention of elias wrt to this from the codebase, maybe some reference to this PR should be enough on `IntegerList`

**Comment by shekhirin** [comment]
> i'd just remove any mention of elias wrt to this from the codebase

yeah, already done, only snapshots and benches left

image

**Comment by gakonst** [comment]
Just confirming that we want this and we'll release alongside the next alpha which breaks the DB.

**Comment by shekhirin** [comment]
@gakonst correct. This PR can already be reviewed, because it'll get merged into `breaking-changes` branch.

**Review by onbjerg — APPROVED** [review]
lgtm

**Pull request merged**